File: from_python/claude_client.py
"""
Клиент для работы с Claude API.

Этот модуль инкапсулирует логику взаимодействия с Anthropic Claude API,
включая отправку запросов и обработку ответов.
"""

import os
import traceback
import time
from typing import Dict, Tuple, Optional, List, Any

from anthropic import Anthropic, APIError, APIConnectionError, RateLimitError, AuthenticationError

from constants import (
    CLAUDE_MODEL,
    MAX_TOKENS,
    OUTPUT_FORMAT_DEFAULT,
    VALID_OUTPUT_FORMATS,
    MAX_REPLY_LOG_LENGTH,
    ERROR_INVALID_API_KEY,
    ERROR_RATE_LIMIT,
    ERROR_CONNECTION,
    ERROR_API_KEY_NOT_SET,
    ERROR_API_KEY_NOT_FOUND,
    HTTP_TOO_MANY_REQUESTS,
    HTTP_INTERNAL_SERVER_ERROR,
    HTTP_SERVICE_UNAVAILABLE
)
from prompts import get_system_prompt, get_user_message, SPEC_END_MARKER
from logger import get_logger
from mcp_tools import MCPTools

logger = get_logger(__name__)


class ClaudeClient:
    """
    Клиент для взаимодействия с Claude API.
    
    Attributes:
        client: Экземпляр Anthropic клиента
        api_key: API ключ для аутентификации
    """
    
    def __init__(self, api_key: Optional[str] = None, enable_mcp: bool = True):
        """
        Инициализирует Claude клиент.

        Args:
            api_key: API ключ Anthropic. Если None, берется из переменной окружения
            enable_mcp: Включить MCP инструменты (по умолчанию True)

        Raises:
            ValueError: Если API ключ не найден
        """
        self.api_key = api_key or os.environ.get("ANTHROPIC_API_KEY")

        if not self.api_key:
            logger.error(ERROR_API_KEY_NOT_FOUND)
            raise ValueError(ERROR_API_KEY_NOT_SET)

        logger.info(f"API ключ загружен: {self.api_key[:10]}...{self.api_key[-4:] if len(self.api_key) > 14 else ''}")

        try:
            self.client = Anthropic(api_key=self.api_key)
            logger.info("Anthropic клиент успешно инициализирован")
        except Exception as e:
            logger.error(f"Ошибка инициализации Anthropic клиента: {str(e)}")
            raise

        # Инициализация MCP инструментов
        self.enable_mcp = enable_mcp
        if self.enable_mcp:
            logger.info("MCP инструменты включены")
    
    def validate_output_format(self, output_format: str) -> bool:
        """
        Проверяет валидность формата вывода.
        
        Args:
            output_format: Формат вывода для проверки
            
        Returns:
            True если формат валиден, False иначе
        """
        return output_format in VALID_OUTPUT_FORMATS
    
    def _get_mcp_tools(self, enabled_tools: Optional[List[str]] = None) -> Optional[List[Dict[str, Any]]]:
        """
        Получает список MCP инструментов в формате Anthropic.

        Args:
            enabled_tools: Список имен включенных инструментов. Если None, возвращаются все.

        Returns:
            Список инструментов или None при ошибке
        """
        if not self.enable_mcp:
            return None

        try:
            all_tools = MCPTools.get_tools_definitions()

            # Если enabled_tools не указан или пустой, возвращаем все
            if not enabled_tools:
                logger.info(f"Получено {len(all_tools)} MCP инструментов (все)")
                return all_tools

            # Фильтруем только включенные инструменты
            filtered_tools = [
                tool for tool in all_tools
                if tool['name'] in enabled_tools
            ]

            logger.info(f"Получено {len(filtered_tools)} MCP инструментов (из {len(enabled_tools)} запрошенных)")
            return filtered_tools if filtered_tools else None
        except Exception as e:
            logger.error(f"Ошибка получения MCP инструментов: {e}")
            return None

    def _handle_tool_calls(
        self,
        message_response: Any,
        messages: List[Dict[str, Any]],
        api_params: Dict[str, Any],
        max_iterations: int = 5
    ) -> Tuple[Optional[str], Optional[Dict]]:
        """
        Обрабатывает tool_use блоки от Claude и выполняет цепочки вызовов.

        Args:
            message_response: Ответ от Claude API
            messages: Список сообщений
            api_params: Параметры API запроса
            max_iterations: Максимальное количество итераций tool calls

        Returns:
            Кортеж (финальный_ответ, usage_info)
        """
        iteration = 0
        current_response = message_response
        total_input_tokens = current_response.usage.input_tokens
        total_output_tokens = current_response.usage.output_tokens

        while iteration < max_iterations:
            # Проверяем наличие tool_use в content
            has_tool_use = any(
                block.type == "tool_use"
                for block in current_response.content
            )

            if not has_tool_use:
                # Нет tool_use, возвращаем текстовый ответ
                text_blocks = [
                    block.text for block in current_response.content
                    if hasattr(block, 'text')
                ]
                final_text = "".join(text_blocks)
                usage = {
                    'input_tokens': total_input_tokens,
                    'output_tokens': total_output_tokens
                }
                return final_text, usage

            # Есть tool_use, обрабатываем
            iteration += 1
            logger.info(f"=== Tool call итерация {iteration} ===")

            # Добавляем ответ ассистента с tool_use в историю
            messages.append({
                "role": "assistant",
                "content": current_response.content
            })

            # Собираем результаты всех tool calls
            tool_results = []
            for block in current_response.content:
                if block.type == "tool_use":
                    tool_name = block.name
                    tool_input = block.input
                    tool_use_id = block.id

                    logger.info(f"Вызов инструмента: {tool_name}")
                    logger.info(f"Аргументы: {tool_input}")

                    try:
                        # Вызываем MCP инструмент
                        start_time = time.time()
                        result = MCPTools.call_tool(tool_name, tool_input)
                        elapsed = time.time() - start_time

                        logger.info(f"Результат {tool_name} (за {elapsed:.2f}s): {result[:200]}..." if result and len(result) > 200 else f"Результат {tool_name}: {result}")

                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": tool_use_id,
                            "content": result or "Инструмент не вернул результат"
                        })

                    except Exception as e:
                        logger.error(f"Ошибка выполнения инструмента {tool_name}: {e}")
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": tool_use_id,
                            "content": f"Ошибка: {str(e)}",
                            "is_error": True
                        })

            # Добавляем результаты инструментов
            messages.append({
                "role": "user",
                "content": tool_results
            })

            # Повторный вызов Claude с результатами
            api_params["messages"] = messages
            current_response = self.client.messages.create(**api_params)

            # Обновляем счетчики токенов
            total_input_tokens += current_response.usage.input_tokens
            total_output_tokens += current_response.usage.output_tokens

        # Достигнут лимит итераций
        logger.warning(f"Достигнут лимит итераций tool calls ({max_iterations})")
        text_blocks = [
            block.text for block in current_response.content
            if hasattr(block, 'text')
        ]
        final_text = "".join(text_blocks)
        usage = {
            'input_tokens': total_input_tokens,
            'output_tokens': total_output_tokens
        }
        return final_text, usage

    def send_message(
        self,
        user_message: str,
        output_format: str = OUTPUT_FORMAT_DEFAULT,
        model: str = CLAUDE_MODEL,
        max_tokens: int = MAX_TOKENS,
        spec_mode: bool = False,
        conversation_history: Optional[List[Dict[str, str]]] = None,
        temperature: float = 1.0,
        enabled_tools: Optional[List[str]] = None
    ) -> Tuple[Optional[str], Optional[Dict], int, Optional[Dict]]:
        """
        Отправляет сообщение в Claude API и возвращает ответ.

        Args:
            user_message: Сообщение пользователя
            output_format: Формат вывода ('default', 'json', 'xml')
            model: Модель Claude для использования
            max_tokens: Максимальное количество токенов в ответе
            spec_mode: Режим сбора уточняющих данных (True/False)
            conversation_history: История диалога (список сообщений с role и content)
            temperature: Температура генерации (0.0 - 1.0)
            enabled_tools: Список имен включенных инструментов. Если None, используются все.

        Returns:
            Кортеж из (ответ, ошибка, HTTP код, usage):
            - ответ: Текст ответа от Claude или None при ошибке
            - ошибка: Словарь с описанием ошибки или None при успехе
            - HTTP код: Код статуса HTTP
            - usage: Словарь с информацией о токенах (input_tokens, output_tokens) или None
        """
        try:
            # Валидация формата
            if not self.validate_output_format(output_format):
                logger.warning(f"Неподдерживаемый формат: {output_format}, используется default")
                output_format = OUTPUT_FORMAT_DEFAULT

            # Получаем системный промпт
            try:
                system_prompt = get_system_prompt(output_format, spec_mode)
            except ValueError as e:
                logger.error(f"Ошибка получения системного промпта: {str(e)}")
                return None, {'error': str(e)}, HTTP_INTERNAL_SERVER_ERROR, None

            # Получаем чистое сообщение пользователя
            clean_user_message = get_user_message(user_message)

            # Формируем массив сообщений с историей
            messages = []

            # Добавляем историю диалога (если есть)
            if conversation_history:
                for msg in conversation_history:
                    if msg.get('role') in ('user', 'assistant') and msg.get('content'):
                        messages.append({
                            "role": msg['role'],
                            "content": msg['content']
                        })

            # Добавляем текущее сообщение пользователя
            messages.append({
                "role": "user",
                "content": clean_user_message
            })

            # Получаем MCP инструменты с учетом фильтра
            tools = self._get_mcp_tools(enabled_tools)

            # Формируем параметры запроса
            api_params = {
                "model": model,
                "max_tokens": max_tokens,
                "system": system_prompt,
                "messages": messages,
                "temperature": temperature
            }

            # Добавляем инструменты если есть
            if tools:
                api_params["tools"] = tools

            # Добавляем stop_sequences для spec mode
            if spec_mode: api_params["stop_sequences"] = [SPEC_END_MARKER]

            # === Детальное логирование параметров API запроса ===
            logger.info("=== Отправка запроса к Claude API ===")
            logger.info(f"Модель: {model}")
            logger.info(f"Max tokens: {max_tokens}")
            logger.info(f"Temperature: {temperature}")
            if tools:
                logger.info(f"MCP инструменты: {[t['name'] for t in tools]}")
            # Логируем system prompt (первые 200 символов)
            system_preview = system_prompt[:200] + "..." if len(system_prompt) > 200 else system_prompt
            logger.info(f"System prompt ({len(system_prompt)} символов): \"{system_preview}\"")

            # Логируем сообщения
            logger.info(f"Сообщения ({len(messages)} шт.):")
            for i, msg in enumerate(messages, 1):
                content = msg['content']
                if isinstance(content, str):
                    content_preview = content[:100] + "..." if len(content) > 100 else content
                    # Убираем переносы строк для компактности
                    content_preview = content_preview.replace('\n', ' ').replace('\r', '')
                    logger.info(f"  [{i}] {msg['role']} ({len(content)} символов): \"{content_preview}\"")
                else:
                    logger.info(f"  [{i}] {msg['role']} (structured content)")

            if spec_mode:
                logger.info(f"Stop sequences: {api_params.get('stop_sequences', [])}")
            logger.info("=====================================")

            # Отправляем запрос к API
            request_start = time.time()
            message = self.client.messages.create(**api_params)
            request_time = time.time() - request_start

            # Проверяем наличие tool_use
            has_tool_use = any(
                block.type == "tool_use"
                for block in message.content
                if hasattr(block, 'type')
            )

            if has_tool_use and tools:
                # Обрабатываем tool calls
                logger.info("Ответ содержит tool_use, начинаем обработку...")
                raw_reply, usage = self._handle_tool_calls(message, messages, api_params)
                logger.info(f"Tool calls завершены за {time.time() - request_start:.2f}s")
            else:
                # Обычный текстовый ответ
                raw_reply = message.content[0].text
                usage = {
                    'input_tokens': message.usage.input_tokens,
                    'output_tokens': message.usage.output_tokens
                }

            logger.info(f"Сырой ответ от Claude: {raw_reply[:MAX_REPLY_LOG_LENGTH]}...")
            logger.info(f"Использовано токенов: input={usage['input_tokens']}, output={usage['output_tokens']}")
            logger.info(f"Общее время запроса: {time.time() - request_start:.2f}s")

            return raw_reply, None, 200, usage

        except AuthenticationError as e:
            logger.error(f"Ошибка аутентификации API: {e}\n{traceback.format_exc()}")
            return None, {'error': ERROR_INVALID_API_KEY}, HTTP_INTERNAL_SERVER_ERROR, None

        except RateLimitError as e:
            logger.error(f"Превышен лимит запросов: {e}\n{traceback.format_exc()}")
            return None, {'error': ERROR_RATE_LIMIT}, HTTP_TOO_MANY_REQUESTS, None

        except APIConnectionError as e:
            logger.error(f"Ошибка соединения с API: {e}\n{traceback.format_exc()}")
            return None, {'error': ERROR_CONNECTION}, HTTP_SERVICE_UNAVAILABLE, None

        except APIError as e:
            logger.error(f"Ошибка Claude API: {e}\n{traceback.format_exc()}")
            return None, {'error': f'Ошибка Claude API: {str(e)}'}, HTTP_INTERNAL_SERVER_ERROR, None

        except Exception as e:
            logger.error(f"Неожиданная ошибка: {e}\n{traceback.format_exc()}")
            return None, {'error': f'Ошибка сервера: {str(e)}'}, HTTP_INTERNAL_SERVER_ERROR, None
    
    def is_api_key_configured(self) -> bool:
        return bool(self.api_key)

File: app/src/main/kotlin/com/claude/agent/services/ClaudeClient.kt
package com.claude.agent.services

import com.claude.agent.config.AppConfig
import com.claude.agent.config.ClaudeConfig
import com.claude.agent.config.ErrorMessages
import com.claude.agent.config.SPEC_END_MARKER
import com.claude.agent.models.Message
import com.claude.agent.models.SystemPrompts
import com.claude.agent.models.TokenUsage
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import org.slf4j.LoggerFactory
import kotlinx.serialization.json.add

/**
 * Клиент для работы с Claude API (Anthropic).
 *
 * Аналог claude_client.py из Python-версии.
 * Использует ktor-client для HTTP запросов к Anthropic API.
 */
class ClaudeClient(
    private val httpClient: HttpClient,
    private val mcpTools: MCPTools
) {
    private val logger = LoggerFactory.getLogger(ClaudeClient::class.java)
    private val apiKey = AppConfig.anthropicApiKey
    private val apiUrl = "https://api.anthropic.com/v1/messages"

    companion object {
        private const val MAX_TOOL_ITERATIONS = 5
    }

    val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)

    /**
     * Отправляет сообщение в Claude API и возвращает ответ.
     *
     * @param clientIp IP-адрес клиента для автоматического определения местоположения в MCP инструментах
     * @param userLocation Координаты пользователя из браузера (если доступны)
     * @return Triple(reply, usage, error)
     */
    suspend fun sendMessage(
        userMessage: String,
        outputFormat: String = "default",
        model: String = ClaudeConfig.MODEL,
        maxTokens: Int = ClaudeConfig.MAX_TOKENS,
        specMode: Boolean = false,
        conversationHistory: List<Message> = emptyList(),
        temperature: Double = 1.0,
        enabledTools: List<String> = emptyList(),
        clientIp: String? = null,
        userLocation: com.claude.agent.models.UserLocation? = null
    ): Triple<String?, TokenUsage?, String?> {
        try {
            // Формируем системный промпт
            val systemPrompt = SystemPrompts.getSystemPrompt(outputFormat = outputFormat, specMode = specMode, enabledTools = enabledTools)
            val cleanUserMessage = SystemPrompts.getUserMessage(userMessage)

            // Формируем массив сообщений с историей
            val messages = buildMessages(conversationHistory, cleanUserMessage)

            // Получаем remote MCP серверы
            val remoteMcp = mcpTools.getRemoteMCP()

            // Получаем инструменты (включая remote MCP toolsets)
            val tools = getFilteredTools(enabledTools, remoteMcp)

            // Формируем запрос
            val requestBody = buildAnthropicRequest(
                model = model,
                maxTokens = maxTokens,
                systemPrompt = systemPrompt,
                messages = messages,
                temperature = temperature,
                tools = tools,
                remoteMcp = remoteMcp,
                specMode = specMode
            )

            logger.info("=== Отправка запроса к Claude API ===")
            logger.info("Модель: $model")
            logger.info("Max tokens: $maxTokens")
            logger.info("Temperature: $temperature")
            if (tools != null) {
                logger.info("MCP инструменты: ${tools.size} шт.")
            }
            logger.info("Сообщений: ${messages.size}")
            logger.info("=====================================")

            // Отправляем запрос
            val startTime = System.currentTimeMillis()
            val response: HttpResponse = httpClient.post(apiUrl) {
                header("x-api-key", apiKey)
                header("anthropic-version", "2023-06-01")
                // Добавляем beta header для поддержки MCP connector
                if (remoteMcp.isNotEmpty()) {
                    header("anthropic-beta", "mcp-client-2025-11-20")
                }
                contentType(ContentType.Application.Json)
                setBody(requestBody)
            }

            if (!response.status.isSuccess()) {
                val errorBody = response.bodyAsText()
                logger.error("Ошибка API: ${response.status}, body: $errorBody")
                return Triple(null, null, "Ошибка Claude API: ${response.status}")
            }

            val responseBody = response.body<JsonObject>()
            val elapsed = System.currentTimeMillis() - startTime

            // Обрабатываем ответ (с поддержкой tool_use)
            val (finalReply, totalUsage) = handleResponse(
                responseBody,
                messages,
                model,
                maxTokens,
                systemPrompt,
                temperature,
                tools,
                remoteMcp,
                clientIp,
                userLocation
            )

            logger.info("Ответ получен за ${elapsed}ms")
            logger.info("Использовано токенов: input=${totalUsage?.input_tokens}, output=${totalUsage?.output_tokens}")

            return Triple(finalReply, totalUsage, null)

        } catch (e: Exception) {
            logger.error("Ошибка sendMessage: ${e.message}", e)
            return Triple(null, null, "Ошибка сервера: ${e.message}")
        }
    }

    /**
     * Обрабатывает ответ от Claude API (включая tool_use цепочки).
     */
    private suspend fun handleResponse(
        responseBody: JsonObject,
        initialMessages: List<JsonObject>,
        model: String,
        maxTokens: Int,
        systemPrompt: String,
        temperature: Double,
        tools: JsonArray?,
        remoteMcp: JsonArray,
        clientIp: String?,
        userLocation: com.claude.agent.models.UserLocation?
    ): Pair<String, TokenUsage> {
        var currentResponse = responseBody
        val messages = initialMessages.toMutableList()
        var totalInputTokens = currentResponse["usage"]?.jsonObject?.get("input_tokens")?.jsonPrimitive?.int ?: 0
        var totalOutputTokens = currentResponse["usage"]?.jsonObject?.get("output_tokens")?.jsonPrimitive?.int ?: 0

        var iteration = 0

        while (iteration < MAX_TOOL_ITERATIONS) {
            val content = currentResponse["content"]?.jsonArray ?: JsonArray(emptyList())

            // Проверяем наличие tool_use
            val hasToolUse = content.any { block ->
                block.jsonObject["type"]?.jsonPrimitive?.content == "tool_use"
            }

            if (!hasToolUse) {
                // Финальный текстовый ответ
                val textBlocks = content.mapNotNull { block ->
                    if (block.jsonObject["type"]?.jsonPrimitive?.content == "text") {
                        block.jsonObject["text"]?.jsonPrimitive?.content
                    } else null
                }
                val finalText = textBlocks.joinToString("")
                val usage = TokenUsage(totalInputTokens, totalOutputTokens)
                return Pair(finalText, usage)
            }

            // Есть tool_use - обрабатываем
            iteration++
            logger.info("=== Tool call итерация $iteration ===")

            // Добавляем ответ ассистента в историю
            messages.add(JsonObject(mapOf(
                "role" to JsonPrimitive("assistant"),
                "content" to content
            )))

            // Выполняем все tool calls
            val toolResults = mutableListOf<JsonObject>()
            for (block in content) {
                val blockObj = block.jsonObject
                if (blockObj["type"]?.jsonPrimitive?.content == "tool_use") {
                    val toolName = blockObj["name"]?.jsonPrimitive?.content ?: continue
                    val toolInput = blockObj["input"]?.jsonObject ?: JsonObject(emptyMap())
                    val toolUseId = blockObj["id"]?.jsonPrimitive?.content ?: continue

                    logger.info("Вызов инструмента: $toolName")

                    val result = try {
                        mcpTools.callTool(toolName, toolInput, clientIp, userLocation)
                    } catch (e: Exception) {
                        logger.error("Ошибка выполнения $toolName: ${e.message}")
                        """{"error": "${e.message}"}"""
                    }

                    toolResults.add(JsonObject(mapOf(
                        "type" to JsonPrimitive("tool_result"),
                        "tool_use_id" to JsonPrimitive(toolUseId),
                        "content" to JsonPrimitive(result)
                    )))
                }
            }

            // Добавляем результаты инструментов
            messages.add(JsonObject(mapOf(
                "role" to JsonPrimitive("user"),
                "content" to JsonArray(toolResults)
            )))

            // Повторный запрос к Claude
            val requestBody = buildAnthropicRequest(
                model = model,
                maxTokens = maxTokens,
                systemPrompt = systemPrompt,
                messages = messages,
                temperature = temperature,
                tools = tools,
                remoteMcp = remoteMcp,
                specMode = false
            )

            val response: HttpResponse = httpClient.post(apiUrl) {
                header("x-api-key", apiKey)
                header("anthropic-version", "2023-06-01")
                // Добавляем beta header для поддержки MCP connector
                if (remoteMcp.isNotEmpty()) {
                    header("anthropic-beta", "mcp-client-2025-11-20")
                }
                contentType(ContentType.Application.Json)
                setBody(requestBody)
            }

            currentResponse = response.body<JsonObject>()
            totalInputTokens += currentResponse["usage"]?.jsonObject?.get("input_tokens")?.jsonPrimitive?.int ?: 0
            totalOutputTokens += currentResponse["usage"]?.jsonObject?.get("output_tokens")?.jsonPrimitive?.int ?: 0
        }

        // Лимит итераций достигнут
        logger.warn("Достигнут лимит tool call итераций ($MAX_TOOL_ITERATIONS)")
        val content = currentResponse["content"]?.jsonArray ?: JsonArray(emptyList())
        val textBlocks = content.mapNotNull { block ->
            if (block.jsonObject["type"]?.jsonPrimitive?.content == "text") {
                block.jsonObject["text"]?.jsonPrimitive?.content
            } else null
        }
        val finalText = textBlocks.joinToString("")
        val usage = TokenUsage(totalInputTokens, totalOutputTokens)
        return Pair(finalText, usage)
    }

    private fun buildMessages(history: List<Message>, userMessage: String): List<JsonObject> {
        val messages = mutableListOf<JsonObject>()

        // Добавляем историю
        for (msg in history) {
            if (msg.role in listOf("user", "assistant") && msg.content.isNotBlank()) {
                messages.add(JsonObject(mapOf(
                    "role" to JsonPrimitive(msg.role),
                    "content" to JsonPrimitive(msg.content)
                )))
            }
        }

        // Добавляем текущее сообщение
        messages.add(JsonObject(mapOf(
            "role" to JsonPrimitive("user"),
            "content" to JsonPrimitive(userMessage)
        )))

        return messages
    }

    private fun getFilteredTools(enabledTools: List<String>, remoteMcp: JsonArray): JsonArray? {
        val allTools = mcpTools.getToolsDefinitions(enabledTools)
        val filtered = allTools.filter { it.name in enabledTools }

        val elements = buildList {
            // Если есть remote MCP серверы, добавляем mcp_toolset для каждого
            if (remoteMcp.isNotEmpty()) {
                remoteMcp.forEach { serverElement ->
                    val serverObj = serverElement.jsonObject
                    val serverName = serverObj["name"]?.jsonPrimitive?.content
                    if (serverName != null) {
                        add(JsonObject(mapOf(
                            "type" to JsonPrimitive("mcp_toolset"),
                            "mcp_server_name" to JsonPrimitive(serverName)
                        )))
                    }
                }
            }

            // Добавляем локальные инструменты (если есть)
            filtered.forEach { tool ->
                add(JsonObject(mapOf(
                    "name" to JsonPrimitive(tool.name),
                    "description" to JsonPrimitive(tool.description),
                    "input_schema" to tool.input_schema
                )))
            }
        }

        return if (elements.isNotEmpty()) JsonArray(elements) else null
    }

    private fun buildAnthropicRequest(
        model: String,
        maxTokens: Int,
        systemPrompt: String,
        messages: List<JsonObject>,
        temperature: Double,
        remoteMcp: JsonArray,
        tools: JsonArray?,
        specMode: Boolean
    ): JsonObject {
        val params = mutableMapOf(
            "model" to JsonPrimitive(model),
            "max_tokens" to JsonPrimitive(maxTokens),
            "system" to JsonPrimitive(systemPrompt),
            "messages" to JsonArray(messages),
            "temperature" to JsonPrimitive(temperature),
        )

        if (remoteMcp.isNotEmpty()) {
            params["mcp_servers"] = remoteMcp
        }

        if (tools != null) {
            params["tools"] = tools
        }

        if (specMode) {
            params["stop_sequences"] = JsonArray(listOf(JsonPrimitive(SPEC_END_MARKER)))
        }

        return JsonObject(params)
    }

    fun isApiKeyConfigured(): Boolean = apiKey.isNotBlank()
}
